Design/Workbench/design-projectsaving.md

```diff
@@ -0,0 +1,173 @@
+﻿
+Project Saving for Mantid Workbench
+==============================
+Motivation
+----------
+In MantidPlot it was possible to save all currently open windows and all workspaces, and then be able to load then back in the exact(-ish) state that they were in previously. To achieve this goal again in Mantid Workbench, the problem will be split into two where window serialization and saving and therefore reopening will be separate in implementation, from saving workspaces.
+
+The reason for the split is so that it can be handled in a more abstract manor. Instead of thinking of the task as one great big issue it can be split into the two smaller problems and then split further from there, later on in design.
+
+This document exists because Project Saving is a non-trivial challenge to tackle when it comes to ideas with saving and loading interfaces created in both Python and C++ as well as saving and loading of workspaces. This issue should be fairly simple to understand, simply users should be able to save their projects and load them back in, to at least the standard that MantidPlot was able to achieve.
+
+It should as a requirement allow people to share projects across platforms and seamlessly by zipping and emailing or transferring by different methods. This however should not come in the way of pure functionality.
+
+The Saver/Serializer should take from interfaces it's dictionary from the output and if possible turn it into JSON and then also be able to transfer it back in the exact same form it was recieved when loading the data back in, the rest should be handled by indidual developers implementing and maintaining interfaces. This has the added benefit of allowing someone in the future to rip out the JSON parts and replace it with whatever saver they would like to use.
+
+Changes to MantidWorkbench needed
+---------------------------------
+- MainWindow should track all of the open windows (as well as if they are serializable or not) similar to MantidPlot
+- File->Save Project needs to be added
+- File->Load Project needs to be added
+- A GUI for picking where to Save
+
+GUI and integration initial ideas
+---------------------------------
+It is true that this document is mostly focused on the actual implementation of saving the project and they intricacies of how that would work. However, it is important that this is discussed to some extent before hand. The intial thought process is that you would be able to do both a basic SaveAll and Save Advanced. SaveAll will run through a pretty simple Save everything as long as it's not overwriting workspaces. Save Advanced would allow the users to Select workspaces and windows individually, also adding an option as to whether or not they want to overwrite the workspaces that may or may be saved which is so at worst case users can save everything from scratch.
+
```

Comment:
Fo you see the `Save Advance` as a per-save setting. As in the user would be prompted every time to state exactly what items they wanted to save? Is there any thoughts towards a general persisted set of advanced settings?

Reply:
I hadn't given that much thought but it should be possible with QSettings/Serialization I presume.